File: AdcKi/Week#1/UVA10855.py
# UVA 10855 - Rotated squares
# 2019.09.17
# Prog by AdcKi
# TLE

import sys
import math

def rotateSquare(n, Square):
  for i in range(n):
    Square[i] = list(Square[i])
  for x in range(int(math.ceil(n / 2.0))):
    for y in range(int(math.floor(n / 2.0))):
      swap = Square[x][y]
      Square[x][y] = Square[n - 1 - y][x]
      Square[n - 1 - y][x] = Square[n - 1 - x][n - 1 - y]
      Square[n - 1 - x][n - 1 - y] = Square[y][n - 1 - x]
      Square[y][n - 1 - x] = swap

  for i in range(n):
    Square[i] = ''.join(Square[i])
  return Square

def countSSinBS(n, m, BS, SS):
  cnt = 0
  for i in range(n - m + 1):
    for j in range(n - m + 1):
      flag = False
      for k in range(m):
        for l in range(m):
          if BS[i + k][j + l] != SS[k][l]:
            flag = True
            break
        if flag == True:
          break
      if flag == False:
        cnt += 1
  return cnt

while True:
  try: (n, m) = map(int, sys.stdin.readline().split())
  except: break
  if n == 0 and m == 0: break
  BS = []
  SS = []
  for i in range(n):
    str = sys.stdin.readline()
    BS.append(str)
  for i in range(m):
    str = sys.stdin.readline()
    SS.append(str)
  for rotateCount in range(4):
    print(countSSinBS(n, m, BS, SS), end = ' ')
    SS = rotateSquare(m, SS)
  print()

File: AdcKi/Week#1/UVA10855.cpp
// UVA 10855 - Rotated square
// 2019.09.17
// Prog by AdcKi
// AC

#include<bits/stdc++.h>

using namespace std;

const int MAX_SIZE = 1000 + 5;

char BS[MAX_SIZE][MAX_SIZE], SS[MAX_SIZE][MAX_SIZE];

bool isSS(int y, int x, int m) {
	for (int i = 0; i < m ; i++) {
		for (int j = 0; j < m; j++) {
			if (BS[y + i][x + j] != SS[i][j]) {
				return false;
			}
		}
	}
	return true;
}

int main() {
	cin.tie(NULL);
	ios_base::sync_with_stdio(false);
	while (true) {
		int n, m;
		cin >> n >> m;
		if (n == 0 && m == 0) return 0;
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				cin >> BS[i][j];
			}
		}
		for (int i = 0; i < m; i++) {
			for (int j = 0; j < m; j++) {
				cin >> SS[i][j];
			}
		}
		for (int l = 0; l < 4; l++) {
			int cnt = 0;
			for (int i = 0; i < n - m + 1; i++) {
				for (int j = 0; j < n - m + 1; j++) {
					if (isSS(i, j, m)) cnt++;
				}
			}
			cout << cnt;
			if (l == 3) break;
			cout << " ";
			for (int i = 0; i < ceil(m / 2.0); i++) {
				for (int j = 0; j < floor(m / 2.0); j++) {
					char swap = SS[i][j];
					SS[i][j] = SS[m - 1 - j][i];
					SS[m - 1 - j][i] = SS[m - 1 - i][m - 1 - j];
					SS[m - 1 - i][m - 1 - j] = SS[j][m - 1 - i];
					SS[j][m - 1 - i] = swap;
				}
			}
		}
		cout << "\n";
	}
	return 0;
}
